### PR TITLE
Adding a method (and impls) to get a snapshot of the ClusterMap

### PR DESCRIPTION
**Pull request by vgkholla**
Added method and impls to get a snapshot of the state of the `ClusterMap` at the time a request was made. This will help with debugging unexpected cluster states.

Will add support in frontend and server to fetch this data via http/custom requests.

Will add tests in a follow up PR

### CODE REVIEW

File: ambry-store/src/test/java/com.github.ambry.store/StoreTestUtils.java [hunk 1]
```diff
@@ -109,6 +110,11 @@ public boolean isSealed() {
       return isSealed;
     }
 
+    @Override
+    public JSONObject getSnapshot() {
+      return new JSONObject();
```
Comment (jsjtzyy) [inline ambry-store/src/test/java/com.github.ambry.store/StoreTestUtils.java]: just curious why here returns new `JSONObject`, however, in previous test, `null` is directly returned?

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/StaticClusterManager.java [hunk 2]
```diff
@@ -498,6 +502,29 @@ public void onReplicaEvent(ReplicaId replicaId, ReplicaEventType event) {
...
+    snapshot.put(IMPLEMENTATION, StaticClusterManager.class.getName());
+    snapshot.put(CLUSTER_NAME, partitionLayout.getClusterName());
+    snapshot.put(TIMESTAMP_MS, SystemTime.getInstance().milliseconds());
+    JSONArray datacentersJsonArray = new JSONArray();
+    hardwareLayout.getDatacenters().forEach(dc -> {
+      JSONObject data = new JSONObject();
+      data.put(DATACENTER_NAME, dc.getName());
+      data.put(DATACENTER_ID, dc.getId());
+      JSONArray datanodesInDc = new JSONArray();
+      dc.getDataNodes().forEach(mockDataNodeId -> datanodesInDc.put(mockDataNodeId.getSnapshot()));
```
Comment (lightningrob) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/StaticClusterManager.java]: Minor comment: should this be called mockDataNodeId given it's not a test class?  (Maybe was copied code.)

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplica.java [hunk 3]
```diff
@@ -103,6 +105,32 @@ public boolean isSealed() {
...
+    JSONObject snapshot = new JSONObject();
+    DataNodeId dataNodeId = getDataNodeId();
+    snapshot.put(REPLICA_NODE, dataNodeId.getHostname() + ":" + dataNodeId.getPort());
+    snapshot.put(REPLICA_PARTITION, getPartitionId().toPathString());
+    snapshot.put(REPLICA_DISK, getDiskId().getMountPath());
+    snapshot.put(REPLICA_PATH, getReplicaPath());
+    snapshot.put(CAPACITY_BYTES, getCapacityInBytes());
+    snapshot.put(REPLICA_WRITE_STATE, isSealed() ? PartitionState.READ_ONLY.name() : PartitionState.READ_WRITE.name());
+    String replicaLiveness = UP;
+    if (dataNodeId.getState() == HardwareState.UNAVAILABLE) {
```
Comment (cgtz) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplica.java]: is this the precedence order followed by the router/ambryrequests?

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java [hunk 4]
```diff
@@ -316,6 +321,29 @@ public void onReplicaEvent(ReplicaId replicaId, ReplicaEventType event) {
...
+    snapshot.put(IMPLEMENTATION, HelixClusterManager.class.getName());
+    snapshot.put(CLUSTER_NAME, clusterName);
+    snapshot.put(TIMESTAMP_MS, SystemTime.getInstance().milliseconds());
+    JSONArray datacentersJsonArray = new JSONArray();
+    dcIdToDcName.forEach((dcId, dcName) -> {
+      JSONObject data = new JSONObject();
+      data.put(DATACENTER_NAME, dcName);
+      data.put(DATACENTER_ID, dcId);
+      JSONArray datanodesInDc = new JSONArray();
+      dcToNodes.get(dcName).forEach(mockDataNodeId -> datanodesInDc.put(mockDataNodeId.getSnapshot()));
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java]: Renamed

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/Replica.java [hunk 5]
```diff
@@ -100,6 +102,28 @@ public boolean isSealed() {
...
+    snapshot.put(CAPACITY_BYTES, getCapacityInBytes());
+    snapshot.put(REPLICA_WRITE_STATE, isSealed() ? PartitionState.READ_ONLY.name() : PartitionState.READ_WRITE.name());
+    String replicaLiveness = UP;
+    if (dataNodeId.getState() == HardwareState.UNAVAILABLE) {
+      replicaLiveness = NODE_DOWN;
+    } else if (disk.getState() == HardwareState.UNAVAILABLE) {
+      replicaLiveness = DISK_DOWN;
+    } else if (isStopped) {
+      replicaLiveness = REPLICA_STOPPED;
+    }
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/Replica.java]: Added another `if`

File: ambry-store/src/test/java/com.github.ambry.store/StoreTestUtils.java (hunk 1 above)
Comment (vgkholla) [reply ambry-store/src/test/java/com.github.ambry.store/StoreTestUtils.java]: No preference. Made this return `null` too.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/StaticClusterManager.java (hunk 2 above)
Comment (vgkholla) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/StaticClusterManager.java]: Fixed

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDatanode.java [hunk 6]
```diff
@@ -146,6 +153,34 @@ public long getXid() {
     return xid;
   }
 
+  @Override
+  public JSONObject getSnapshot() {
+    JSONObject snapshot = new JSONObject();
+    snapshot.put(DATA_NODE_HOSTNAME, getHostname());
+    snapshot.put(DATA_NODE_DATACENTER, getDatacenterName());
+    snapshot.put(DATA_NODE_SSL_ENABLED_DATACENTERS, new JSONArray(sslEnabledDataCenters));
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDatanode.java]: It should be as long as the config is similar - but this class is making no assumptions

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplica.java (hunk 3 above)
Comment (vgkholla) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplica.java]: It looks to be so - it seems to be precedence order of `isDown()`

**Comment by vgkholla** [comment]
@cgtz @jsjtzyy @lightningrob addressed comments. I'll add tests in a follow up PR (they can get very lengthy)

**Pull request merged**